opentracing/index.js

```diff
@@ -0,0 +1,165 @@
+  }
+
+  // TODO: Align with other agents about TEXT/BINARY format
+  // TODO: Maybe wait implementing the binary format until: https://github.com/w3c/trace-context/tree/master/trace_context#binary-format
+  // - On the other hand, as we're not putting this is a special location, we might not have to wait on Trace Context - similar to how we don't wait for the HTTP definition when using a custom header prefix
+  _inject (spanContext, format, carrier) {
+    switch (format) {
+      case Tracer.FORMAT_TEXT_MAP:
+      case Tracer.FORMAT_HTTP_HEADERS:
+        carrier['elastic-apm-traceparent'] = spanContext.toString()
+        break
+      case Tracer.FORMAT_BINARY:
+        carrier.buffer = spanContext._elasticContext.buffer
+        break
+    }
+  }
+
+  _extract (format, carrier) {
+    let ctx
+
+    switch (format) {
+      case Tracer.FORMAT_TEXT_MAP:
+      case Tracer.FORMAT_HTTP_HEADERS:
+        ctx = ElasticTraceContext.fromString(carrier['elastic-apm-traceparent'])
+        break
+      case Tracer.FORMAT_BINARY:
+        ctx = new ElasticTraceContext(carrier.buffer)
+        break
+    }
+
+    return ctx ? new SpanContext(ctx) : null
+  }
+}
+
+class Span extends ot.Span {
+  constructor (tracer, span) {
+    super()
+    this.__tracer = tracer
+    this.__context = null
+    this._span = span
+    this._isTransaction = span instanceof Transaction
+  }
+
+  _context () {
+    return this.__context
+      ? this.__context
+      : (this.__context = new SpanContext(this._span.context))
+  }
+
+  _tracer () {
+    return this.__tracer
+  }
+
+  _setOperationName (name) {
+    this._span.name = name
+  }
+
+  _addTags (tags) {
+    // If we delete the tags below, we should make sure to not mutate the tags
+    // object
+    tags = Object.assign({}, tags)
+
+    if (this._isTransaction) {
+      if (tags.result) {
+        this._span.result = tags.result
+        delete tags.result // TODO: Should this be deleted?
+      } else if (tags['http.status_code']) {
+        this._span.result = `HTTP ${String(tags['http.status_code'])[0]}xx`
+        delete tags['http.status_code'] // TODO: Should this be deleted?
+      }
+    }
+
+    if (tags.type) {
+      this._span.type = tags.type
+      delete tags.type // TODO: Should this be deleted?
+    }
+
+    const id = tags['user.id']
+    const username = tags['user.username']
+    const email = tags['user.email']
+    if (id || username || email) {
+      // TODO: Shold we differentiate if these tags are set on a span or a transaction?
+      this._agent.setUserContext({ id, username, email })
+      // TODO: Should these be deleted?
+      delete tags['user.id']
+      delete tags['user.username']
+      delete tags['user.email']
+    }
+
+    if (Object.keys(tags).length === 0) return
+
+    if (!this._isTransaction) return this._agent.logger.info('Elastic APM OpenTracing: Could not set tags on span') // The Elastic APM agent currently doesn't support span tags
```

Comment:
nit: this.__tracer.agent.logger

Reply:
We since added support for tags on spans, so I've removed this line entirely 😄